astroid/context.py

```diff
@@ -42,7 +42,11 @@ class InferenceContext:
 
     max_inferred = 100
 
-    def __init__(self, path=None, nodes_inferred=None):
+    def __init__(
+        self,
+        path: set[tuple[NodeNG, str | None]] | None = None,
```

Comment:
Where do you base this on?

Reply:
```
        :type: set(tuple(NodeNG, optional(str)))

        Path of visited nodes and their lookupname

        Currently this key is ``(node, context.lookupname)``
```

Reply:
It's very likely that this is indeed `SuccesfulInferenceResult` instead of `NodeNG`. Let's revert this and keep it without the typing for now. The PR is really useful anyway!

Reply:
I changed `NodeNG` to `Any`. Even with that hole, this a pretty specific type to verify.

I have a feeling that this pattern is going to come up all over the place, where a docstring says it's `NodeNG`, but in reality it's `NodeNG | Proxy | Base | Etc`. Being able to verify the compound types without having to untangle all the node-like types would be easier.

Reply:
I don't really understand your last comment.

I agree it would be good to untangle where the docstring annotation is incorrect but I prefer to do so in separate PRs as checking that is so involved. If we save that change for a small follow-up PR we can focus on getting that right while we can now already merge the `context: InferenceContext | None` stuff.